File: back/app/schemas/Item.py
from pydantic import BaseModel, RootModel
from typing import Dict, Literal, Set, Union

class Gold(BaseModel):
    base: int
    purchasable: bool
    total: int
    sell: int


class Stat(BaseModel):
    name: str
    kind : Literal["flat", "percentage"] 
    value : float | int
    #Makes hashable to use in the set
    class Config:
        frozen = True

class Effects(RootModel):
    root: Dict[str, Union[int, float]]

class Item(BaseModel):
    name: str
    plaintext: str
    image: str 
    imageUrl: str
    gold: Gold
    tags: Set[str] 
    stats: Set[Stat] 
    effect: Effects
    id: int
    description: str

File: front/src/app/interfaces/Item.ts
export interface Gold {
  base: number;
  purchaseable: boolean;
  total: number;
  sell: number;
}

export enum StatKind {
  Flat,
  Percentage,
}

export enum EffectKind {
  effect1,
  effect2,
}

export interface Stat {
  name: string;
  kind: StatKind;
  value: number;
}

export interface Effect {
  name: string;
  value: number;
  kind: EffectKind;
}

export interface Item {
  name: string;
  gold: Gold;
  description: string;
  stats: Stat[];
  tags: string[];
  effects: Effect[];
  img: string;
  id: number;
  hasHdImage?: boolean;
}

export enum FilterItemSortField {
  PRICE = "Price",
  NAME = "Name",
}

export enum FilterSortOrder {
  ASC = "asc",
  DESC = "desc",
}
